common/scanner.h

```diff
@@ -1,294 +1,444 @@
-      advance(lexer);
-      lexer->result_symbol = RIGHT_QUOTED_STRING_DELIM;
-      return scan_right_quoted_string_delimiter(lexer);
-    } else if (in_string && valid_symbols[STRING_DELIM] && lexer->lookahead == '"') {
-      advance(lexer);
-      in_string = false;
-      lexer->result_symbol = STRING_DELIM;
-      return true;
+#define STRING_CLEAR(vec)                                                      \
+    {                                                                          \
+        (vec).len = 0;                                                         \
+        memset((vec).data, 0, (vec).cap * sizeof(char));                       \
     }
 
-    while (iswspace(lexer->lookahead)) {
-      skip(lexer);
-    }
+enum TokenType {
+    COMMENT,
+    LEFT_QUOTED_STRING_DELIM,
+    RIGHT_QUOTED_STRING_DELIM,
+    STRING_DELIM,
+    LINE_NUMBER_DIRECTIVE,
+    NULL_CHARACTER
+};
 
-    if (!in_string && lexer->lookahead == '#' && lexer->get_column(lexer) == 0) {
-      advance(lexer);
+typedef struct {
+    uint32_t cap;
+    uint32_t len;
+    char *data;
+} String;
 
-      while (lexer->lookahead == ' ' || lexer->lookahead == '\t') advance(lexer);
+String string_new() {
+    return (String){.cap = 16, .len = 0, .data = calloc(1, sizeof(char) * 17)};
+}
 
-      if (!iswdigit(lexer->lookahead)) return false;
-      while (iswdigit(lexer->lookahead)) advance(lexer);
+typedef struct {
+    bool in_string;
+    String quoted_string_id;
+} Scanner;
 
-      while (lexer->lookahead == ' ' || lexer->lookahead == '\t') advance(lexer);
+Scanner *scanner_new() {
+    Scanner *scanner = (Scanner *)calloc(1, sizeof(Scanner));
+    scanner->in_string = false;
+    scanner->quoted_string_id = string_new();
+    return scanner;
+}
+
+static inline void advance(TSLexer *lexer) { lexer->advance(lexer, false); }
 
-      if (lexer->lookahead != '"') return false;
-      while (lexer->lookahead != '\n' && lexer->lookahead != '\r' && lexer->lookahead != '"') advance(lexer);
-      if (lexer->lookahead != '"') return false;
+static inline void skip(TSLexer *lexer) { lexer->advance(lexer, true); }
 
-      while (lexer->lookahead != '\n' && lexer->lookahead != '\r') advance(lexer);
+unsigned serialize(Scanner *scanner, char *buffer) {
+    return 0;
+    size_t size = scanner->quoted_string_id.len;
 
-      lexer->result_symbol = LINE_NUMBER_DIRECTIVE;
-      return true;
-    } else if (!in_string && lexer->lookahead == '(') {
-      advance(lexer);
-      lexer->result_symbol = COMMENT;
-      return scan_comment(lexer);
-    } else if (!in_string && valid_symbols[STRING_DELIM] && lexer->lookahead == '"') {
-      advance(lexer);
-      in_string = true;
-      lexer->result_symbol = STRING_DELIM;
-      return true;
-    } else if (valid_symbols[NULL_CHARACTER] && lexer->lookahead == '\0') {
-      return !lexer->eof(lexer);
+    if (size + 1 > TREE_SITTER_SERIALIZATION_BUFFER_SIZE) {
+        return 0;
```

Comment:
Isn't it better to crash instead? Returning 0 will result in weird behaviour,

Reply:
No, it's common in most scanners to just truncate extra data, crashing wouldn't be ideal - same principle applies in python, bash, etc

Reply:
tree-sitter-cpp is [crashing](https://github.com/tree-sitter/tree-sitter-cpp/blob/0e7b7a02b6074859b51c1973eb6a8275b3315b1d/src/scanner.c#L129-L131). It would be possible to only save the start of `quoted_string_id`, but that complicates the implementation, and in practice `quoted_string_id` will never be that long.

Reply:
Hmm yeah, I guess crashing could be fine here too